casper/src/main/resources/MultiSigRevVault.rho

```diff
@@ -0,0 +1,263 @@
+      }
+    } |
+
+    contract MultiSigRevVault(@"deployerAuthKey", deployerId, ret) = {
+      new DeployerIdOps(`rho:rchain:deployerId:ops`),
+          revAddrCh,
+          deployerPubKeyBytesCh in {
+        DeployerIdOps!("pubKeyBytes", *deployerId, *deployerPubKeyBytesCh) |
+        for (@deployerPubKeyBytes <- deployerPubKeyBytesCh) {
+          _fromPublicKey!(deployerPubKeyBytes, *ret)
+        }
+      }
+    } |
+
+    contract _fromPublicKey(@publicKey, ret) = {
+      new RevAddress(`rho:rev:address`), revAddrCh in {
+        RevAddress!("fromPublicKey", publicKey, *revAddrCh)|
+        for (@revAddress <- revAddrCh) {
+          @AuthKey!("make", (*_multiSigRevVault, revAddress), *ret)
+        }
+      }
+    } |
+
+    // MultiSig
+    // Returns (false, errorMsg) or (true, (multiSigVault, revAddress, revVault))
+    contract MultiSigRevVault(@"create", @publicKeys, @unsealers, @quorumSize, ret) = {
+      if (quorumSize > publicKeys.length() + unsealers.length() or
+          quorumSize <= 0) {
+        ret!((false, "invalid quorum size"))
+      } else {
+        new unf, authKeyCh, authPubKeysCh, RevAddress(`rho:rev:address`), revAddrCh in {
+          RevAddress!("fromUnforgeable", *unf, *revAddrCh) |
+          @RevVault!("unforgeableAuthKey", *unf, *authKeyCh) |
+          @ListOps!("map", publicKeys, *_fromPublicKey, *authPubKeysCh) |
+          for (@revVaultAuthKey <- authKeyCh; @authPubKeys <- authPubKeysCh; @revAddr <- revAddrCh) {
+            new multiSig, nonceStore, nonceMapStore, revVaultCh, verify in {
+              @RevVault!("findOrCreate", revAddr, *revVaultCh) |
+              for (@maybeRevVault <- revVaultCh) {
+                match maybeRevVault {
+                  (false, _) => { ret!(maybeRevVault) }
+                  (true, revVault) => {
+                    ret!((true, (*multiSig, revAddr, revVault))) |
+                    // Nonce starts at zero
+                    @(*multiSig, *nonceStore)!(0) |
+
+                    // Map from (revAddress, nonce) to
+                    //   (true, confirmationSet)
+                    //   (false, "transaction complete")
+                    @(*multiSig, *nonceMapStore)!({}) |
+
+                    contract multiSig(@"balance", ret) = {
+                      @revVault!("balance", *ret)
+                    } |
+
+                    contract verify(@auth, idxAuthPubKeysCh, unsealedCh) = {
+                      new unsealerTest in {
+                        // Check whether auth is in the list of authPubKeys
+                        @ListOps!("indexOf", authPubKeys, auth, *idxAuthPubKeysCh) |
+                        // Check whether auth is a sealed tuple
+                        @ListOps!("partialFold", unsealers, (-1, Nil), *unsealerTest, *unsealedCh) |
+                        contract unsealerTest(head, @(prevPos, _), result) = {
+                          new retH in {
+                            head!(auth, *retH) |
```

Comment:
It's difficult to read because `auth` is not an `auth` at all in this case but a sealed tuple.
Maybe add a comment?

Reply:
It is a value demonstrating that the request is authorized.  It is not necessarily an auth*Key*, which is why I changed the name.  The contract verify is for internal use only; there are comments on the public-facing API where auth is provided on lines 127 and 174.